many sd-boot/bootctl fixes, and a new "boot counting" concept, for automatic fallback to older kernels on persistent failures

poettering opened the pull request:
Yupp, a lot of patches in a single PR, but most are rather simple and the stuff at the end is mostly docs...

In order to grok the concepts introduced here it might make sense to read through the last commit first which contains docs explaining how this all fits together.

vcaputo commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,197 @@
+For unattended systems such as servers, IoT/embedded devices or appliances an
+automatic logic of reverting back to previous versions of the OS or kernel in
+case system boots consistently fail is important. systemd provides support for
+this in various of its components. When used together they provide a complete
```
s/various/a variety/

vcaputo commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,197 @@
+* The
+  [`systemd-boot(7)`](https://www.freedesktop.org/software/systemd/man/systemd-boot.html)
+  boot loader optionally maintains a per-boot-loader-entry counter that is
+  decreased by one on each attempt to boot the entry, priorizing entries that
```
s/priorizing/prioritizing/

vcaputo commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,197 @@
+
+* The
+  [`systemd-boot-check.service(8)`](https://www.freedesktop.org/software/systemd/man/systemd-boot-check.service.html)
+  service is a simple health check tool that determines whether the boot
```
s/whether/wether/

vcaputo commented:
@poettering I can see the appeal of overloading the entry filenames to contain the attempts/remaining counters, but I think it would probably be better to leave the entries alone.  What if someone wanted to have the directory or fs containing the entries be non-writable outside of privileged update windows when they are manipulated?

Maybe have a separate directory containing the per-entry counters, and use symlinks named after the entry files.  Make the link target contain the counters rather than actually pointing anywhere meaningful in the filesystem.

This also would eliminate the issue of entries unintentionally containing counters in their name, if that was a possibility (I didn't examine if the naming of those entries was entirely user-defined).

evverx commented:
This pull request **introduces 1 alert** when merging 1d8536b58288edfd40e242f82ee88c7f205fae9a into ad4bc3352285f467f4ffa03c3171b19fa0a8758d - [view on LGTM.com](https://lgtm.com/projects/g/systemd/systemd/rev/pr-73c20e629e4a86a9f2f7ea34bc73781f4b3007d8)

**new alerts:**

* 1 for Comparison result is always the same

---

*Comment posted by [LGTM.com](https://lgtm.com)*

dnicolodi commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,197 @@
+  service automatically marks a boot loader entry for which boot counting as
+  mentioned above is enabled, as "good" when a boot has been determined to be
+  successful, thus turning off boot counting for it again.
+
```
One comma is missing: "marks a boot loader **entry,** for which boot counting as mentioned above is enabled, as "good" when a boot has been determined to be successful".

dnicolodi commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,197 @@
+   transaction.
+
+8. The `boot-complete.target` unit is ordered after and pulls in various units
+   that shall are required to succeed for the boot process to be considered
```
s/shall//

medhefgo commented:
Also, I feel like turning boot counting off after the first successful boot is the wrong way to go. The whole point of this is to be failsafe, no? And booting could go wrong later again...

poettering commented:
> @poettering I can see the appeal of overloading the entry filenames to contain the attempts/remaining counters, but I think it would probably be better to leave the entries alone. What if someone wanted to have the directory or fs containing the entries be non-writable outside of privileged update windows when they are manipulated?
> 
> Maybe have a separate directory containing the per-entry counters, and use symlinks named after the entry files. Make the link target contain the counters rather than actually pointing anywhere meaningful in the filesystem.

sd-boot operates on the ESP, i.e. FAT. FAT doesn't have symlinks. And even if we loosen the requirement for FAT/ESP one day we should still stick to operations that are implementable on FAT.

We need to be able to manipulate the counters from two sides: the boot loader, before we start a boot process. And the OS, after we completed a boot process. (And of course, also during update, to install the image and set an initial count on it.) These time windows are all very brief and well defined I'd say, on their own.

> 
> This also would eliminate the issue of entries unintentionally containing counters in their name, if that was a possibility (I didn't examine if the naming of those entries was entirely user-defined).

I thought about that, and if it indeed turns out to be a problem I figure we should simply make the whole logic something one can turn off, so that the "+" stuff looses its magic effect. After all the whole design is put together so that a boot loader that implements the boot loader spec but doesn't do boot counting will just work, but ignore the counters.

poettering commented:
> I generally like this but I really disagree on using rename to take track of tries. We already distrust EFI firmwares (for good reason), so using EFI variables for this is out of the question. 

The distrust in EFI variables is probably more caused by the quality of memory chips backing them, i.e. it's doubt about the hardware more than the software.

> But then we also cannot rely on them implementing renames in an atomic fashion. What if we are really lucky and lose power while the EFI is touching the fs? We could end up losing the boot config file or it could end up with garbage. And if it happens to be the only one the user has (for whatever reason) we wouldn't even have anything to boot anymore.

Well, that leaves very little wiggle room... We only really have the ESP and the vars as storage space.

medhefgo commented:
It's fine putting them in ESP, just use a separate file instead of the loader entries. You could put them in /loader/tries/* for example.

poettering commented:
> It's fine putting them in ESP, just use a separate file instead of the loader entries. You could put them in /loader/tries/* for example.

I think that's more risky actually, as writing files means actually allocating/releasing new blocks usually, and that is hard to do atomically (and also involves allocation algorithms, and there's a good chance firmware will fuck that up). Renaming otoh is very likely implementable atomically even on crappy, simpler file systems... I mean, we have no golden bullet, we can just pick the least bad option from a pool of crappy ones. I am pretty sure renaming is safer than writing file contents...

What I like about the rename approach is firstly that the operation it implements has a chance of being atomic and reliable, but also that this means the counting info is directly associated with the entry, and the lifecycle of this data always bound together. You cannot remove the entry without also removing the associated counting data, and that's a really good thing. If you split the counting data into a separate file you need to be careful to always explicitly lifecycle that counting data along with the entry as they are not directly and implicitly attached to each other.

poettering commented:
> Also, I feel like turning boot counting off after the first successful boot is the wrong way to go. The whole point of this is to be failsafe, no? And booting could go wrong later again...

Hmm, so downgrade attacks are a major problem in computer security, in particular in cryptography. After we verified that a kernel works sufficiently well for us to mark the entry good, we should really stick to it I think. There are two opposing concepts here I figure: "don't make yourself vulnerable to downgrade attacks" and "downgrade if upgrade is fucked". I think we can only marry these two if we say we never downgrade once an upgrade worked at least once.

poettering commented:
> With regards to the rename atomicity argument, the last time I checked actual on-disk atomicity of rename was not something guaranteed by most if not all filesystems. POSIX only speaks to runtime atomicity with regards to visibility of the replaced name to other processes. It's a common misconception to believe this equates durably atomic renames across unclean mounts of the filesystem, but that's not really the case IIRC.

Yeah, that's why I say there's a "chance" it's atomic. But you are right, there's no guarantee, and shitty code will always make this hard. But it's taking the least bad from a pile of crappy options.

medhefgo commented:
I can see the point on downgrading attacks.

Another thing: I feel like counting tries is overkill if you just want to make sure upgrades work. If the new kernel+initrd don't work, retrying would be futile, no? So making this into a boolean signaling if this booted successful once should be enough.

medhefgo commented:
I just had a crazy idea: in theory you could store this information in the file mtime... :D

poettering commented:
> Another thing: I feel like counting tries is overkill if you just want to make sure upgrades work. If the new kernel+initrd don't work, retrying would be futile, no? So making this into a boolean signaling if this booted successful once should be enough.

Well, you never know what people do: maybe they turn off the machine during boot (think: some random appliance handled by random people, or even a desktop PC with an encrypted harddisk and a password prompt in the initrd). There might be many reasons why a kernel/os image doesn't boot properly and only a subset of them have anything to do with the image itself. Hence I think there should be some room for configuration here. If deployments are in particular volatile environments it might make sense to initialize the counter to something high to allow a larger number of tries before giving up and reverting back. If deployments are in very well defined, very deterministic and supervised environments then it might make sense to initialize it to something low, maybe even 1, to only offer one attempt before reverting back.

evverx commented on src/boot/boot-complete.c:
```diff
@@ -0,0 +1,444 @@
+
+        r = make_good(prefix, suffix, &good);
+        if (r < 0)
+                return log_oom(),
```
I thought the alert was a false positive until I noticed this comma. I guess a semicolon should be here instead.

poettering replied:
ouch! I kept scratching my head about this warning, and couldn't find what was wrong! Thanks!

poettering commented on src/boot/efi/boot.c:
```diff
@@ -41,6 +41,11 @@ typedef struct {
         EFI_STATUS (*call)(VOID);
         BOOLEAN no_autoselect;
         BOOLEAN non_unique;
+        UINTN tries_done;
```
well, but that would suggest the counter could actually go negative, but it can't... We use (UINTN) -1 as a special way to indicate one particular exceptional case (and we do this similar elsewhere in the tree), hence I don't think we should really change the type altogether....

poettering commented:
I force pushed a new version with all reported typos fixed!

boucman commented:
I know non-EFI system are not well handled by systemd, but having a systemd-provided point for "system is good" would be very interesting for embedded systems, and embedded/IoT systems don't have EFI...

Would it be possible to split the various units into a generic bootloader-independant part and a EFI-specific part, the same way the wait-online in network-manager-independant ?

This way we could andd u-boot or barebox specific implementations easily...

poettering commented:
> I know non-EFI system are not well handled by systemd, but having a systemd-provided point for "system is good" would be very interesting for embedded systems, and embedded/IoT systems don't have EFI...
> 
> Would it be possible to split the various units into a generic bootloader-independant part and a EFI-specific part, the same way the wait-online in network-manager-independant ?

That's actually excactly what this is already doing, and exactly this is even explicitly mentioned in the markdown document explaining how this all fits together, at the second to last commit of this PR. Please have a look.
 
> This way we could andd u-boot or barebox specific implementations easily...

Yes, absolutely, the "boot-complete.target" is supposed to be generic, and usable with other boot loaders and other "boot ok" schemes.

boucman commented:
right, my bad...

I'll reread the whole PR in more details. I was kinda rushed when I first read it.

sorry

boucman commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,203 @@
+    off.
+
+12. On the following boot (and all subsequent boots after that) the entry is
+    now seen with boot counting turned off, no further renaming takes place.
```
Maybe consider reseting the counter instead of disabling boot-counting entirely.

I understand the idea of "if it booted once properly, the image is correct and it won't magically stop working" but I have actually seen that happen in real life. 

Suppose you have two completely separate (i.e different rootfs) boot entries. One is the normal operation one, the other is a recovery system that can do only one thing : reformat and reinstall the normal partition.

The normal partition might become bad for various reasons on a completely autonomous system. Bad blocks, bad handling of disk full, bad recovery on incorrect shutdown... Those kind of things are very rare, but on autonomous systems without any sysadmin they can happen and, statistically, they do.

At this point, having the boot counter still active provides a simple "parachute" for a complete reinstall, and that's very welcome.

poettering replied:
@medhefgo also suggested this, see above. and also see my comments on that. Besides the downgrade-attack thing I also think there's value in putting boundaries on the time where we keep changing the $BOOT file system. With the implemented approach, if an update cycle is clean we'll update $BOOT three times: once to install the new kernel, once on the next boot to count down, and once after it completed to mark it OK. After that $BOOT isn't touched anymore, regardless how many cycles follow, until the next update. I think that's much better behaviour than continuing to count forerver.

I mean, the current scheme makes its changes to the ESP, which is VFAT, and hence has less than ideal integrity behaviour. If we had something more reliable we'd use it, but it's what we have to work with, but we should still optimize the write cycles, and I think O(1) instead of O(n) write cycles per update is highly preferable.

But dunno. we can certainly look into making other schemes workable if there's really the need, once the initial support is in.

boucman replied:
That's no big deal, I'll have to reimplement that part of the logic for u-boot anyway...

that was more a case of "i've been there before" but there are many use cases and only one can be the default...

boucman commented on units/boot-complete.target:
```diff
@@ -0,0 +1,14 @@
+Description=Boot Completion Check
+Documentation=man:systemd.special(7)
+Requires=sysinit.target
+After=sysinit.target
```
I think a very common use-case will be to unconditionnaly use boot-complete.target.

maybe add

[Install]
RequiredBy=default.target

so it is easy to do ? (of course, i'm not asking to enbale the target by default, just that the section be here for ease of use)

poettering replied:
What would the benefit of this be? If there's nothing ordered after the target that depends on it then all you get by this is some extra log output about a couple of failing services?

boucman replied:
maybe it's a missunderstanding on my side...

IIUC the way you see things, a user that wants to activate boot counting will enable systemd-boot-complete.service which will pull boot-complete.target right ?

It would make more sense to me to enable/disable boot-complete.target and have systemd-boot-complete.service have a WantedBy on it... This way the way to enable boot counting is bootloader-neutral.

poettering replied:
> IIUC the way you see things, a user that wants to activate boot counting will enable systemd-boot-complete.service which will pull boot-complete.target right ?

no. not quite. systemd-boot-complete.service is not supposed to be enabled. Instead, systemd-boot-complete-generator automatically pulls it into the initial transaction if it detects that boot counting is enabled.

Basically, to make use of all of this on your uefi system it is entirely sufficient to do `echo > /etc/kernel/tries`. The next kernel you install then will have boot counting enabled, and all will work fully automatically, systemd-boot will make use of it, and the generator ensures thes userspace side of this runs too.

> It would make more sense to me to enable/disable boot-complete.target and have systemd-boot-complete.service have a WantedBy on it... This way the way to enable boot counting is bootloader-neutral.

So my thinking was that there might be other reasons you'd want a boot complete check in place, not just boot counting stuff. For example, let's say you design a system where worker systems shall request a job to run from some central system if they booted up correctly, and only then. For usecases like that it might make sense to reuse boot-complete.target, even though this has very little to do with boot counting.

boucman replied:
well I don't use UEFI (well, I mean... I don't configure UEFI) so I don't really know the whole systemd-boot stuff. I don't really understand why you use a generator for boot counting instead of enabling boot counting statically or using a command-line parameter... Is it to have a way to enable it through UEFI variables ?

Anyway, if that's the idea I missed it in the doc. I have never seen a use-case where boot counting is dynamically disable. For all my use-cases it could be a compile option. (yes that's stupid for other reasons, but that would work for me)

Because I don't completely see what you are trying to do I probably miss the big picture, but I would like to avoid having to write a generator to replace yours for the embedded use-case, kowing it would just recreate a link every time that could be described with an [Install] section

As for enabling boot-complete.target vs enabling a systemd-boot specific service... my main grip is that's systemd-boot specific. so one more thing that i need to redo in a bootloader specific way...

Thing is... UEFI is an x86 only thing... (there was a move to implement it on ARM, but this seems to be dead by now, afaict) so we need to reimplement systemd-boot and its internal logic, which is not really documented. If there is a way to simplify that part, it's always better for us...

keszybz replied:
A generator is preferable because it is "configurationless". Essentially, you decide boot-counting=yes/no before a kernel package is installed, and the rest happens automatically. We *could* ask the user to provide a proper switch somewhere at the kernel command-line, but setting the kernel command-line is fragile, and could get out of sync, etc. By auto-detecting if we need to do this, we make it very easy for the user, and also easier for the system installer. I don't think making a generator is so bad. Usually they are very short and simple, can be written in any language, and they are also easy to test.

poettering replied:
> well I don't use UEFI (well, I mean... I don't configure UEFI) so I don't really know the whole systemd-boot stuff. I don't really understand why you use a generator for boot counting instead of enabling boot counting statically or using a command-line parameter... Is it to have a way to enable it through UEFI variables ?

The only reason to have the generator is to make this less fragile. Instead of having to explicitly turn on boot counting in the boot loader and then also in systemd, it's entirely sufficient if you turn this on in the boot loader in a completely generic fashion, and systemd will pick it up.

My goal here was to make the coupling as lose but automatic as we can: everything gets detected automatically if the boot loader supports it, and all the files on disk actually remain compatible with boot loaders that do not support it.

Now, on UEFI we have the luxury of UEFI variables which can be checked very early on, before we mount any file systems. Very likely other boot systems do not have this luxury, and need to store their state exclusively in the file system. In that case a generator can't really work, and hence this PR cannot be used as 1:1 blueprint for that case. In that case it's probably advisible to instead make their systemd-boot-complete.service counterpart something you explicitly enable.

boucman replied:
well, it's more of the other way round on arm... there is usually a dedicated NAND sector for booloader variables, and we try to avoid relying on the filesystem at all... boot-counting is usually done in those variables rather than in a FAT partition which needs to be doubled for robustness purpose.

but I see where you are going. My use-case is usually simpler, since in the embedded world we don't care as much about automated deployment in the way datacenters do. For me kernel command line is good enough since we never deactivate boot counting.

poettering commented on doc/AUTOMATIC_BOOT_ASSESSMENT.md:
```diff
@@ -0,0 +1,203 @@
+  completed successfuly. When enabled it becomes an indirect dependency of
+  `systemd-boot-complete.service` (by means of `boot-complete.target`, see
+  below), ensuring that the boot will not be considered successful if there are
+  any failed services.
```
We generally prefix stuff units that run code provided by us with "systemd-", and this service does invoke C code we hacked up, hence it carries that prefix.

keszybz commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -63,6 +64,13 @@ if ! [[ ${BOOT_OPTIONS[*]} ]]; then
 fi
 
+if [[ -f /etc/kernel/tries ]]; then
+    read -r TRIES < /etc/kernel/tries
```
There's no space after `<` or `>` elsewhere in this file, please remove it here too.

keszybz commented on src/kernel-install/90-loaderentry.install:
```diff
@@ -63,6 +64,13 @@ if ! [[ ${BOOT_OPTIONS[*]} ]]; then
 
+if [[ -f /etc/kernel/tries ]]; then
+    read -r TRIES < /etc/kernel/tries
+    LOADER_ENTRY="$BOOT_ROOT/loader/entries/$MACHINE_ID-$KERNEL_VERSION+$TRIES.conf"
```
There needs to be a check if `TRIES` is a valid integer (and set at all).

keszybz commented on meson.build:
```diff
@@ -1726,6 +1726,14 @@ if conf.get('ENABLE_EFI') == 1 and conf.get('HAVE_BLKID') == 1
                    install : true,
                    install_dir : rootlibexecdir)
+
+        executable('systemd-boot-complete-generator',
```
I think this should be 'systemd-boot-update-counter`, which is much more self-explanatory.

poettering replied:
well, but it actually doesn't "update" it, but actually removes it... It's "sd-boot" that updates it.

Independently of this: the fact that this is a counter appears too much of an implementation detail to me. I mean if the user does "systemctl list-units" and sees "systemd-boot-complete.service", then that is a friendly high-level indication I think, and whether this is a bit or a counter or some kind of other concept shouldn't matter, I think.

keszybz replied:
I still find the name confusing. The generator and the service are named after the time when the service is called, not after it's purpose. Let's consider the service first. It is described as "Mark the Current Boot Loader Entry as Good", so it should be called something like `systemd-mark-boot-entry-as-good.service`, or maybe `systemd-bless-boot-entry.service` (a bit perlesque, but appropriate), or something. Then the generator would be called `systemd-bless-boot-entry-generator` or whatever. The problem I have with current names is that everything is so confusing because everything is named so similarly: `boot-complete.target`, `systemd-boot-complete-generator`, `systemd-boot-complete.service`, even though they serve different purposes and have different levels of generality.

keszybz commented on src/boot/boot-complete-generator.c:
```diff
@@ -0,0 +1,75 @@
+#include "virt.h"
+
+/* This generator pulls in systemd-boot-complete.service if the "LoaderBootCountPath" EFI variable is set, i.e. the
+ * system boots up with boot counting in effect, which means we should mark the boot is "good" if we manage to boot up
```
is "good" → as "good" ?

keszybz commented on src/boot/boot-complete-generator.c:
```diff
@@ -0,0 +1,75 @@
+#include "util.h"
+#include "virt.h"
+
+/* This generator pulls in systemd-boot-complete.service if the "LoaderBootCountPath" EFI variable is set, i.e. the
```
pulls ... into *what* ?

keszybz commented on src/boot/boot-complete-generator.c:
```diff
@@ -0,0 +1,75 @@
+
+        /* We pull this in from basic.target so that it ends up in all "regular" boot ups, but not in rescue.target or
+         * even emergency.target. */
+        p = strjoina(arg_dest, "/" SPECIAL_BASIC_TARGET ".wants/systemd-boot-complete.service");
```
No, this is cannot be. This needs to be hooked up late into boot, after all the checks and whatnot have happened. Marking a boot as successful early in the boot is pointless.

poettering replied:
(misunderstanding, as discussed below)

keszybz commented on mkosi.build:
```diff
@@ -90,3 +90,8 @@ cat > "$DESTDIR"/etc/issue <<EOF
+# Manually update the boot loader from the one we just built
+mkdir -p "$DESTDIR"/boot/efi/EFI/systemd "$DESTDIR"/boot/efi/EFI/BOOT
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/boot/efi/EFI/systemd/systemd-bootx64.efi
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/boot/efi/EFI/BOOT/bootx64.efi
```
Hmm, are those paths (with /boot/efi) always correct?

keszybz commented on mkosi.build:
```diff
@@ -90,3 +90,8 @@ cat > "$DESTDIR"/etc/issue <<EOF
+# Manually update the boot loader from the one we just built
+mkdir -p "$DESTDIR"/boot/efi/EFI/systemd "$DESTDIR"/boot/efi/EFI/BOOT
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/boot/efi/EFI/systemd/systemd-bootx64.efi
+cp "$DESTDIR"/usr/lib/systemd/boot/efi/systemd-bootx64.efi "$DESTDIR"/boot/efi/EFI/BOOT/bootx64.efi
```
Hmm, are those paths (with /boot/efi) always correct?

poettering replied:
in the mkosi image we build they are...

keszybz commented on src/boot/efi/boot.c:
```diff
@@ -28,7 +28,7 @@ enum loader_type {
 
 typedef struct {
-        CHAR16 *file;
+        CHAR16 *id; /* The identifier for this entry (not that this id is not necessarily unique though!) */
```
Too many "not"s.

poettering commented:
> My major gripe is with the positioning of of the boot counter update service. IIUC, it will be called very early, when it actually needs to be called very late.

The boot counter update service `systemd-boot-complete.service` is ordered after `sysinit.target` by default, as well as `boot-complete.target`. The latter is supposed to be the extension point which people can use to configure precisely how early or late they want the boot counter update to happen. If they don't order anything else before it, yes it will run pretty early on. But they can order anything they like before it, for example:

1. let's say httpd.service and mysql.service is extremely imporant to them. In that case they can add Requires= pointing from `boot-complete.target` towards them (through RequiredBy=), and the marking will be delayed until httpd and mysql are known to have started up successfuly.

2. They can also enable our own "systemd-boot-check.service", which is mostly an example but can be useful too. It is ordered by default after multi-user.target and graphical.target, i.e very late and it checks for any failed services.  They can also hack up their own services like this, possible with simple scripts, and order them properly to move the marking as early or late into the boot process as they wish.

So, by default we'll mark things as good early on, but people have total freedom there, and can move it somewhere in the middle of the boot process (which option 1 above would do) or very later into it (which option 2 would do).

keszybz commented:
Ah, OK. I was looking at the generator, and the generator pulls this into basic.target, and the ordering wrt. boot-complete.target is done in the unit file, so I missed that. The ordering is fine in that case.

poettering commented on man/systemd-boot.xml:
```diff
@@ -336,11 +336,68 @@
+    specific entry continously fails. Any boot loader entry file and unified kernel image file that contains a
+    <literal>+</literal> followed by one or two numbers (if two they need to be separated by a <literal>-</literal>),
+    before the <filename>.conf</filename> or <filename>.efi</filename> suffix is subject to boot counting: the first of
+    the two numbers ('left tries') is decreased by one on every boot attempt, the second of the two numbers ('done
```
will do tree-wide

poettering commented on man/systemd-boot.xml:
```diff
@@ -336,11 +336,68 @@
+
+    <para>The <citerefentry><refentrytitle>kernel-install</refentrytitle><manvolnum>8</manvolnum></citerefentry> kernel
+    install framework optionally sets the initial 'left tries' counter to the value specified in
+    <filename>/etc/kernel/tries</filename> when a boot loader entry is first created.</para>
```
kernel-install is usually called from the rpm postinst scripts, hence adding cmdline args won't do much good i figure.

poettering commented on man/kernel-install.xml:
```diff
@@ -63,49 +63,62 @@
+
+            <listitem><para><filename>90-loaderentry.install</filename> copies <replaceable>KERNEL-IMAGE</replaceable>
+            to
+            <filename>/boot/<replaceable>MACHINE-ID</replaceable>/<replaceable>KERNEL-VERSION</replaceable>/linux</filename>.
```
No, this one is actually right... Note that there are two different things:

 * $ESP/loader/entries/
 * /boot/<machine-id>/<kernel-version>/

The former is the .conf drop-in of the boot loader spec. The latter is where kernel-install by default places kernels. The latter is actually already adopted by Fedora. The former is not. The latter is used as path regardless of the used boot loader and such.

poettering commented on src/boot/boot-complete.c:
```diff
@@ -0,0 +1,444 @@
+        r = acquire_boot_count_path(&path, &prefix, NULL, &done, &suffix);
+        if (r == -EUNATCH) /* acquire_boot_count_path() won't log on its own for this specific error */
+                return log_error_errno(r, "Not booted with boot counting in effect.");
+        if (r < 0)
```
I was thinking about doing that. The reason why LoaderEntrySelected= is one thing, and LoaderBootCountPath= is another has to do with the effect that the former contains a boot loader entry id, while the latter contains a path to the item that is counted. While the id is generally derived from the path, they are different things. The id doesn't carry the counters (and it shouldn't, so that the same entry gets the same id assigned on each boot regardless of the current state of the counter), and the id might be derived from a .conf boot loader spec entry or from a unified kernel image, or from an otherwise automatically discovered item. (And we might beef up the logic with which we populate the entry menu later, at which point there might be even more source the id might be derived from). All this together means that the mapping between path and id is not bijective and not trivial to generate. Given an id finding the path of the entry is hard, it would involving searching through the ESP (and later possibly other locations), and doing pattern matching...

poettering commented on src/basic/utf8.c:
```diff
@@ -359,6 +359,23 @@ char *utf16_to_utf8(const char16_t *s, size_t length) {
+                return NULL;
+
+        for (i = 0; i < length; i++)
+                n[i] = s[i];
```
i'll implement full conversion now

poettering commented on src/basic/utf8.c:
```diff
@@ -312,15 +312,17 @@ size_t utf8_encode_unichar(char *out_utf8, char32_t g) {
 
+        assert(s);
+
         r = new(char, (length * 4 + 1) / 2 + 1);
```
will fix, and add a comment

poettering commented on man/systemd-boot-complete.service.xml:
```diff
@@ -0,0 +1,115 @@
+      <varlistentry>
+        <term><option>status</option></term>
+
+        <listitem><para>The current status of the boot loader entry file or unified kernel image file is shown. This
```
will rename dont-know → indeterminate.

the "clean" state is determined based on whether the efi boot path var is set or not. i.e. if its not set then the boot loader didn't do boot counting, and i think there's a lot of value in reporting that to the user, hence the separate state for this.

poettering commented:
OK, here's a new version. It addresses everything @keszybz pointed out, modulo what I commented on.

In order to implement correct utf8-to-utf16 conversion i added some non-trivial code to utf8.c. That bit deserves the focus, the rest is mostly just typo fixes and renames. I added some tests for the new utf8 code however, hence there's a good chance i got everything right...

poettering commented on src/basic/utf8.c:
```diff
@@ -314,18 +314,25 @@ size_t utf8_encode_unichar(char *out_utf8, char32_t g) {
+
+        /* Input length is in bytes, i.e. the shortest possible character takes 2 bytes. Each unicode character may
+         * take up to 4 bytes in UTF-8. Let's also account for a trailing NUL byte. */
+        if (length * 2 < length)
```
Not really, no. Only the multiplication by two can overflow. After we multiplied we know the lowest bit is off, hence the increment by one will always work...

Generally I think it's wise to check for overflows individually, rather than trying to be smart and converting them into a single test, simply because this all so fragile and the tests are hard to grok in their full effect

jwrdegoede commented:
I just got pointed here from #9897, where I suggest something similar to the boot-complete.targer this adds. 

More similar the different any-ways, 2 differences are:
1.  I think that how to reach / add the target to the transaction should be left open to various use-cases the generator from this patch-set is one, but we need to keep in mind there might be different ways to get the target into the transaction.
2. #9897 also suggest defining a pre target and allow services with a Before= on that target to block reaching the target till they consider the boot complete (and services responding to reaching this state will need a After= on the pre target).

I think it might be good to first resolve #9897 as it seems there is much interest in this atm and having a generic skeleton for this which isolates the "when is the boot complete" side of the equation from the "tell the bootloader (or whatever) that the boot is complete side and I think something along the lines of a boot-complete.target by itself, independent of the bootcounting stuff, would be useful for this.

keszybz commented on units/systemd-boot-check-no-failures.service.in:
```diff
@@ -0,0 +1,24 @@
+[Unit]
+Description=Check if Any System Units Failed
+Documentation=man:systemd-boot-check-no-failures.service(8)
+After=default.target graphical.target multi-user.target
```
I suggest we accept @jwrdegoede's idea and add a `systemd-boot-complete-pre.target`, and order all checkers after it. Then we can add to the actual systemd.unit that is booted an implicit dependency `Before=systemd-boot-complete-pre.target`. Trying to list all possible targets here cannot work and has the disadvantage that if the admin boot to a different systemd.unit, this service can run early and return bogus results.

That new target could be documented as:
- it is ordered after the main transaction is finished
- checker units can order themselves to be after the main startup phase, but are also free to ignore it, if the state that they are checking is available earlier.

poettering commented:
> Now this needs a rebase again. Maybe this could be an opportunity to submit some of the smaller bits as separate PRs, e.g. the utf8 stuff, bootctl fixes, the generic `boot-complete.target`, and then the boot counting part.

Well, github is really not good at making split ups like this easy. These things all depend on each other, hence all you can do is do one PR with commits 1..x, then another PR with commits 1..x+y, and a third one with 1..x+y+z and so on. and then, as soon as one if them needs rebasing/reworking you have to rebase/rework all the other ones too, which is just messy, and explodes in work...

poettering commented:
>     2\. #9897 also suggest defining a pre target and allow services with a Before= on that target to block reaching the target till they consider the boot complete (and services responding to reaching this state will need a After= on the pre target).

I don't get this. boot-complete.target is already the target that stuff that wants to delay marking of a boot as good shall be ordered before. So what is a "pre" target for that supposed to do on top of that?

boot-complete.target is the main watershed already: stuff that shall be required to reach the "good" state should be ordered before, and stuff that wants to operate on "good" shall be ordered after. What kind of extra milestone do you need beyond that? @keszybz @jwrdegoede can you explain?

(Please have a look at AUTOMATIC_BOOT_ASSESSMENT.md contained in this PR it's supposed to explain these semantics.)

poettering commented:
I force pushed a new version now. Only changes: rebase on current git, and renaming of boot-complete → bless-boot. I must agree I like that name much better indeed. I called it "bless-boot" instead of "bless-boot-entry" though as suggested, as I think short is good, and I think the "entry" doesn't really make things clearer. Hope you agree. No other changes.

I don't understand the "pre" unit concept, hence made no changes regarding that. Would be thankful for an elaboration on that @keszybz @jwrdegoede.

poettering commented:
I have made the UTF-8 patches available as #10173 now too, as requested. Since they are required for this PR they are also included here, hence this PR isn't really anymore digestable really...

poettering commented:
Similar, the bootctl patches are now also available as PR #10175, but they include PR #10173 as they require that code.

I am not sure this really made this all easier, tbh, github can't really model these relationships... 

Anyway, ptal, I think for now this is it.

jwrdegoede commented:
On 25-09-18 14:59, Lennart Poettering wrote:
>     |2\. #9897 also suggest defining a pre target and allow services with a Before= on that target to block reaching the target till they consider the boot complete (and services responding to reaching this state will need a After= on the pre target). |
> 
> I don't get this. boot-complete.target is already the target that stuff that wants to delay marking of a boot as good shall be ordered before. So what is a "pre" target for that supposed to do on top of that?

The idea is that "stuff that wants to delay marking of a boot as good" is ordered before
the "pre" target where as unit files for services which react to the good marking,
e.g. a unit doing:

grub2-set-bootflag boot_success

Are marked after the pre.

The way I see it we need 2 things:

1) A generic target to put services which wait for something to come-up / delay reaching the boot-complete state / target
2) A generic target in which units can be added to its .wants which do something once the boot-complete state is reached

So in my mind 1) is boot-complete-pre.target (which will be purely an ordering target) and 2) is boot-complete.target
into which the units for both 1. and 2. get added to its .wants, with an ordering of before or after the pre.

I hope this explains my reasoning behind this.

poettering commented:
@jwrdegoede hmm, not following. Why do you need two targets for that? Why doesn't one suffice? i.e. order the "is good checkers" *before* boot-complete.target and the "good boot markers" *after* boot-complete.target.

Why isn't this one target sufficient, why do you need two targets? or to ask differently: if you have three targets then there are three phases: the phase before the first target, the phase between the two targets and the phase after the second target. What kind of service do you intend to put in which of these phases? i can't see that we'd need three phases, and hence no need to targets but one should suffice. or what am I missing?

Have you read the markdown doc included in this PR?

poettering commented:
> The idea is to have an ordering target and a target in which to install units which must run once the boot is considered complete. IOW the service calling "grub2-set-bootflag boot_success" would get linked to from /lib/systemd/system/boot-complete.target.wants AFAIK ordering an unit after the target which wants it does nothing and it will start right away anyways. One solution would be to put the link in /lib/systemd/system/multi-user.target.wants with an /after/ boot-complete.target in the unit file, but it feels cleaner to me to have a specific target to put these.

I am not sure that having an explicit target for pulling in the "grub2-set-bootflag" thing and equivalents is really desirable. The reason for that is that I think it would be advisable to make this more automatic, i.e. instead of having to enable a service for this, that hooks itself into that target I think it's much preferable if the service gets automatically pulled in whenever a boot loader is found that requires this. This is how this PR implements this for sd-boot: a unit "generator" is added that runs during early boot, checks whether the necessary EFI variable is set that indicates "sd-boot style boot counting has been enabled for this boot", and then dynamically pulls in the "boot blessing" tool into the initial transaction, and only then. This scheme is a lot more robust, and results in the most minimal boot transactions. Specifically this means that "things just work" without manual enabling of any blessing service, and that the "boot checking" services (which are slow, and possibly messy) are only actually pulled in if really needed, i.e. when the "boot blessing" services really need it.

So, I'd really like to push people towards this model: use a generator and don't enable the blessing service unconditionally. I know that in some cases a generator for this isn't feasible though (as knowing whether boot counting is on might depend on some late mounted file system or so), and in that case there isn't too much other option than to enable the service unconditionally, but in that case the right ordering of the service isn't entirely clear and depends on the case hence adding a generic target for it doesn't sound like a good idea either to me. (i.e. if we can't exactly tell you when this second target runs, then maybe we shouldn't have it, and instead leave it to boot loader writers to either pick default.target or multi-user.target or such)

jwrdegoede commented:
On 26-09-18 10:09, Lennart Poettering wrote:
>     The idea is to have an ordering target and a target in which to install units which must run once the boot is considered complete. IOW the service calling "grub2-set-bootflag boot_success" would get linked to from /lib/systemd/system/boot-complete.target.wants AFAIK ordering an unit after the target which wants it does nothing and it will start right away anyways. One solution would be to put the link in /lib/systemd/system/multi-user.target.wants with an /after/ boot-complete.target in the unit file, but it feels cleaner to me to have a specific target to put these.
> 
> I am not sure that having an explicit target for pulling in the "grub2-set-bootflag" thing and equivalents is really desirable. The reason for that is that I think it would be advisable to make this more automatic, i.e. instead of having to enable a service for this, that hooks itself into that target I think it's much preferable if the service gets automatically pulled in whenever a boot loader is found that requires this. This is how this PR implements this for sd-boot: a unit "generator" is added that runs during early boot, checks whether the necessary EFI variable is set that indicates 
> "sd-boot style boot counting has been enabled for this boot", and then dynamically pulls in the "boot blessing" tool into the initial transaction, and only then. This scheme is a lot more robust, and results in the most minimal boot transactions. Specifically this means that "things just work" without manual enabling of any blessing service, and that the "boot checking" services (which are slow, and possibly messy) are only actually pulled in if really needed, i.e. when the "boot blessing" services really need it.
> 
> So, I'd really like to push people towards this model: use a generator and don't enable the blessing service unconditionally. I know that in some cases a generator for this isn't feasible though (as knowing whether boot counting is on might depend on some late mounted file system or so), and in that case there isn't too much other option than to enable the service unconditionally, but in that case the right ordering of the service isn't entirely clear and depends on the case hence adding a generic target for it doesn't sound like a good idea either to me. (i.e. if we can't exactly tell you 
> when this second target runs, then maybe we shouldn't have it, and instead leave it to boot loader writers to either pick default.target or multi-user.target or such)

If you look at: https://github.com/systemd/systemd/issues/9897

I describe a case there where there are no "blessing services":

"The system-boot-success.target maybe be started by adding it to another target's wants list in the case where system-boot-success-pre.target services are used to wait for the system to become ready; or it may be manually started by some other code."

Specifically in case of a desktop/workstation system the idea
is for the desktop environment to decide when the system is
booted successfully (typically when the DE has seen some
interaction from the user indicating that both input devices
and the GPU work properly).

In this case we really don't want the desktop environment to
need to have knowledge of the bootloader, instead it would
start a special target, e.g. the boot-complete.target from this pr.

And that special target would then have the services in there
which let $bootloader know the boot was successful. Those
services could get added to that target by a generator, or
statically that really is orthogonal to the need of having
a special target, rather then having say the generator put
them in say multi-user.target.

I guess we could have a "blessing service" which waits for
the signal from the DE but that feels roundabout to me if
the DE can simply start the target itself (with a bit of
policykit js to allow this).

keszybz commented:
I'll force push this PR (with no changes) so github sees that some of the commits are already merged.

keszybz commented on man/systemd-boot.xml:
```diff
@@ -337,11 +337,68 @@
+    <literal>+</literal> followed by one or two numbers (if two they need to be separated by a <literal>-</literal>),
+    before the <filename>.conf</filename> or <filename>.efi</filename> suffix is subject to boot counting: the first of
+    the two numbers ('tries left') is decreased by one on every boot attempt, the second of the two numbers ('done
+    tries') is increased by one (if 'tries done' is absent it is considered equivalent to 0). Depending on the current
```
'tries done'

keszybz commented on man/systemd-bless-boot.service.xml:
```diff
@@ -0,0 +1,115 @@
+    <para>Internally, the service operates based on the <varname>LoaderBootCountPath</varname> EFI variable (of the
+    vendor UUID <constant>4a67b082-0a4c-41cf-b6c7-440b29bb8c4</constant>), which is passed from the boot loader to the
+    OS. It contains a file system path (relative to the EFI system partition) of the <ulink
+    url="https://github.com/systemd/systemd/blob/master/docs/BOOT_LOADER_SPECIFICATION.md">Boot Loader
```
systemd.io...

keszybz commented on src/boot/bless-boot-generator.c:
```diff
@@ -0,0 +1,75 @@
+
+/* This generator pulls systemd-bless-boot.service into the initial transaction if the "LoaderBootCountPath" EFI
+ * variable is set, i.e. the system boots up with boot counting in effect, which means we should mark the boot as
+ * "good" if we manage to boot up this far. */
```
"this far" is unclear. The generator runs very early, but the blessing happens very late.

poettering replied:
i changed this to "far enough" now. It's supposed to be vague, since after all it's supposed to open what is meant by "successful boot". But of course, the "this" is confusing indeed, since it's not just vague but potentially referencing something in the same paragraph and that's something we should fix which I hope "far enough" does. Hope that makes sense.

poettering commented:
OK, I rebased it on top of current git and pushed it, so that it applies again. I made all suggested changes. I am taking the liberty to put the green label on it, given that the new changes are only about typo fixes. Thanks for the review!
